spec.bs

```diff
@@ -3363,6 +3386,61 @@ The <dfn for=Navigator method>getInterestGroupAdAuctionData(|configIDL|)</dfn> m
 
 # Reporting # {#reporting}
 
+## Common types and algorithms ## {#reporting-common}
+
+A <dfn>reporting bid source</dfn> is an enum with the following possible values:
+<dl dfn-for="reporting bid source">
+: <dfn>`generate-bid`</dfn>
+:: Bid produced by an invocation of `generateBid()`.
+: <dfn>`additional-bid`</dfn>
+:: Bid provided via [=validate and convert additional bids|the additional bids=] mechanism.
+: <dfn>`bidding-and-auction-services`</dfn>
+:: Bid provided via [=parse and validate server response|Bidding and Auction Services=] mechanism.
+
+</dl>
+
+A <dfn>reporting bid key</dfn> is a [=struct=] with the following [=struct/items=]:
+
+  Note: This [=struct=] exists only to uniquely identify places bids came from, avoiding confusion
+  in cases like bids made by the same interest group in different component auctions, or additional
+  bids reusing names of regular interest groups. Implementations can likely find a more efficient
+  way of achieving the same effect. Note that bids returned when `generateBid()` returns multiple
+  items all share a key, as they share reporting information, too.
+
+<dl dfn-for="reporting bid key">
+  : <dfn>context</dfn>
+  :: The [=reporting context=] corresponding to the component (or single-level) auction the bid

```

Comment:
A little confused by what "context" mean here. Is this field more like a "data" or "reports" storing all kinds of reports (currently only fDO)? And why does it belong to a "reporting bid key"? I thought that the other fields made the key, and this was the reporting data for the key. Currently it reads as [=reporting bid key=] has a [=reporting context=], which is a [=map=] whose key is [=reporting bid key=], which seems to be a loop.

Reply:
Basically the particular (sub)auction this is for --- so that bids from same IGs in different component auctions can be told apart. It's not a loop because these are references, but it is a bit redundant; having everything in the key means you can determine who the winner is by comparing the reporting bid keys.

Reply:
what information of [=reporting context=] can tell bids from same IGs in different component auctions apart? We allow two component auctions from the same component seller, right? It's very unlikely for two [=reporting bid keys=] to be the same, but it's still possible for the current definition IIUC.

Reply:
It's reference equality....

Reply:
ah ok.
The current scheme is:
map (auction_config -> reporting context), where reporting context is a struct [=debug reporting info=]. Is it possible to change it to:
map (auction_config -> [=reporting info (renamed from debug report info) =])?

Reply:
> ah ok. The current scheme is: `map (auction_config -> reporting context), where reporting context is a struct [=debug reporting info=]`. Is it possible to change it to: `map (auction_config -> [=reporting info (renamed from debug report info) =])`?

That wouldn't work as well once Private Aggregation is added since you would need a separate map for private aggregation stuff that will have to be passed everywhere. Having a reporting context means we can stick whatever fields we need for private aggregation into the same struct (and there is a lot of stuff I expect to put there).

...As I mentioned before, we could stick real time reporting there, too, to avoid to need to pass it around, but it's not actually partionned by subauction so there would be extra work merging it (though it's like one extra loop level, so may be worth it for the uniformity). The split is mostly useful for the per-participant P.Agg. metrics I want to add, since I'll need space to stick stuff like how long the component auction's bidding script took to fetch, etc.

> 
> Or say, change [=reporting context=] from a struct to "A map from reporting bid key to bid debug reporting info", without needing the [=debug reporting info=] struct layer.

Same thing here.

Reply:
Oh I see, so PAgg map needs to be in parallel to the [=debug reporting info=] since the two maps have different keys? I was thinking maybe put all PAgg/RTR data can be put inside the [[=bid debug reporting info=]](https://pr-preview.s3.amazonaws.com/morlovich/turtledove/pull/1296.html#bid-debug-reporting-info), but seems not working due to they are keyed by different things, probably?

Reply:
Yeah, and there will probably be more than one thing.